internal/router/router.go

```diff
@@ -148,23 +148,16 @@ type ResolvedURL struct {
 // This function panics if urlStr cannot be parsed as a charm.URL
 // or if it is not fully specified, including user and revision.
 func MustNewResolvedURL(urlStr string, promulgatedRev int) *ResolvedURL {
-	url := charm.MustParseURL(urlStr)
+	url := mustParseURL(urlStr)
 	if url.User == "" || url.Revision == -1 {
 		panic(fmt.Errorf("incomplete url %v", urlStr))
 	}
 	return &ResolvedURL{
-		URL:                 *url.WithChannel(""),
+		URL:                 *url,
 		PromulgatedRevision: promulgatedRev,
 	}
 }
 
-// UserOwnedURL returns the non-promulgated URL for the given id.
-// The returned *charm.URL may be modified freely.
-func (id *ResolvedURL) UserOwnedURL() *charm.URL {
```

Comment:
you've checked that nothing was relying on being able to freely modify the UserOwnedURL?